gtfs-validator/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java

```diff
@@ -0,0 +1,131 @@
+import org.mobilitydata.gtfsvalidator.table.GtfsShapeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStop;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTime;
+import org.mobilitydata.gtfsvalidator.table.GtfsStopTimeTableContainer;
+import org.mobilitydata.gtfsvalidator.table.GtfsTrip;
+import org.mobilitydata.gtfsvalidator.table.GtfsTripTableContainer;
+import org.mobilitydata.gtfsvalidator.notice.StopTooFarFromTripShapeNotice;
+
+import static org.locationtech.spatial4j.context.SpatialContext.GEO;
+
+/**
+ * Validates that a stop is within a distance threshold for a trip shape.
+ * <p>
+ * Generated notices:
+ * * StopTooFarFromTripShapeNotice
+ */
+@GtfsValidator
+public class StopTooFarFromTripShapeValidator extends FileValidator {
+    @Inject GtfsStopTimeTableContainer stopTimeTable;
+
+    @Inject GtfsTripTableContainer tripTable;
+
+    @Inject GtfsShapeTableContainer shapeTable;
+
+    @Inject GtfsStopTableContainer stopTable;
+
+    // Spatial operation buffer values
+    public static final double TRIP_BUFFER_METERS = 100;
+    private static final double TRIP_BUFFER_DEGREES = DistanceUtils.KM_TO_DEG * (TRIP_BUFFER_METERS / 1000.0d);
+
+    @Override
+    public void validate(NoticeContainer noticeContainer) {
+        // Cache for previously tested shape_id and stop_id pairs - no need to test them more than once.
+        final Map<String, List<String>> testedCache = new HashMap<>();
+        // Go through the pair of tripId and the corresponding stop time list one by one.
+        for (Map.Entry<String, List<GtfsStopTime>> tripIdStopTimeListEntry : Multimaps.asMap(stopTimeTable.byTripIdMap()).entrySet()) {
+            final String tripId = tripIdStopTimeListEntry.getKey();
+            final List<GtfsStopTime> stopTimeList = tripIdStopTimeListEntry.getValue();
+            final GtfsTrip trip = tripTable.byTripId(tripId);
+            if (trip == null || stopTimeList == null || stopTimeList.isEmpty() || trip.shapeId() == null) {
+                // This rule only applies when all necessary fields exist.
+                continue;
+            }
+            final List<GtfsShape> shapeList = shapeTable.byShapeId(trip.shapeId());
+            if (shapeList == null || shapeList.isEmpty()) {
+                continue;
+            }
+            // Get the trip shape polygon given the Gtfs shapes data and TRIP_BUFFER_DEGREES.
+            Shape tripShapeGivenThreshold = createTripShapePolygonGivenThreshold(shapeList);
+            // Check if each stop is within the buffer polygon.
+            for (GtfsStopTime stopTime : stopTimeList) {
+                GtfsStop stop = stopTable.byStopId(stopTime.stopId());
+                if (stop == null) {
+                    continue;
+                }
+                if (!(stop.locationType() == GtfsLocationType.STOP) && !(stop.locationType() == GtfsLocationType.BOARDING_AREA)) {
```

Comment:
You could make this easier to read like:
```
if (stop.locationType() != GtfsLocationType.STOP && stop.locationType() != GtfsLocationType.BOARDING_AREA) {
```

Reply:
Thank you for the advice! I have fixed that.